### PR TITLE
CORE-8735: Close publisher after it is not used anymore; change the create resource to return the created resource

### PR DESCRIPTION
**Pull request by yift-r3**
Following the comment in https://github.com/corda/corda-runtime-os/pull/3090#pullrequestreview-1295582063 this pull request will close the publisher in the membership client if the configuration had changed.

It also changes the create resource method to return the created resource...

### CODE REVIEW

File: components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt [hunk 1]
```diff
@@ -132,46 +132,61 @@ class MemberOpsClientImpl @Activate constructor(
...
+                        COMPONENT_HANDLE_NAME,
+                        CONFIG_HANDLE_NAME,
+                        COMPONENT_HANDLE_NAME
+                    )
+                )
                 deactivate("Handling the stop event for component.")
             }
 
             is RegistrationStatusChangeEvent -> {
+                coordinator.closeManagedResources(
```
Comment (yift-r3) [inline components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: @owenstanford  - FYI

File: components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt [hunk 2]
```diff
@@ -78,6 +78,12 @@ class MemberOpsClientImpl @Activate constructor(
 
         const val ASYNC_CLIENT_ID = "membership.ops.async"
 
+        const val PUBLISHER_NAME = "MemberOpsClient.publisher"
+        // for watching the config changes
+        const val CONFIG_HANDLE_NAME = "MemberOpsClient.componentHandle"
+        // for checking the components' health
+        const val COMPONENT_HANDLE_NAME = "MemberOpsClient.configHandle"
```
Comment (dimosr) [inline components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]:
```suggestion
        const val COMPONENT_HANDLE_NAME = "MemberOpsClient.componentHandle"
        // for checking the components' health
        const val CONFIG_HANDLE_NAME = "MemberOpsClient.configHandle"
```

File: components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt [hunk 3]
```diff
@@ -132,46 +132,61 @@ class MemberOpsClientImpl @Activate constructor(
...
             }
 
             is StopEvent -> {
-                componentHandle?.close()
-                configHandle?.close()
+                coordinator.closeManagedResources(
+                    setOf(
+                        COMPONENT_HANDLE_NAME,
+                        CONFIG_HANDLE_NAME,
+                        COMPONENT_HANDLE_NAME
```
Comment (dimosr) [inline components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: is there a reason to have `COMPONENT_HANDLE_NAME` twice here?

Comment (dimosr) [reply components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: is one of them supposed to be the publisher instead?

File: components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt [hunk 4]
```diff
@@ -132,46 +132,61 @@ class MemberOpsClientImpl @Activate constructor(
...
                 deactivate("Handling the stop event for component.")
             }
 
             is RegistrationStatusChangeEvent -> {
+                coordinator.closeManagedResources(
+                    setOf(
+                        PUBLISHER_NAME,
+                        CONFIG_HANDLE_NAME,
+                    )
+                )
```
Comment (dimosr) [inline components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: why is this block needed here? 

Is it handling the case where we get a DOWN event from the configuration service? In that case, should we be adding this into the `else` clause in the `when` block below?

Comment (yift-r3) [reply components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: If it's UP we need to wait for the new configuration to create the publisher again.

Comment (dimosr) [reply components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: hm not sure I fully follow.. `UP` is received when the registration was previously `DOWN` right? So, the publisher would have been closed already (if we move that logic in the else block) or not started yet (if it's the first `UP` signal we receive).

Comment (yift-r3) [reply components/membership/membership-client-impl/src/main/kotlin/net/corda/membership/impl/client/MemberOpsClientImpl.kt]: Yes, I think I got it wrong.

**Review by dimosr — APPROVED** [review]
LGTM - but I think it's worth getting a review/approval from @owenstanford as he raised the original issue and he might have some more context.

**Pull request merged**